src/Data/Barbie/Internal/Product.hs

```diff
@@ -237,3 +239,29 @@ instance
 
   gbuniq x = Rec (K1 (buniq x))
   {-# INLINE gbuniq #-}
+
+
+-- --------------------------------
+-- Instances for base types
+-- --------------------------------
+
+instance ProductB Proxy where
+  bprod _ _ = Proxy
+  {-# INLINE bprod #-}
+
+  buniq _ = Proxy
+  {-# INLINE buniq #-}
+
+instance (ProductB a, ProductB b) => ProductB (Product a b) where
+  bprod (Pair ll lr) (Pair rl rr) = Pair (bprod ll rl) (bprod lr rr)
+  {-# INLINE bprod #-}
+
+  buniq x = Pair (buniq x) (buniq x)
+  {-# INLINE buniq #-}
+
+instance Monoid a => ProductB (Const a) where
```

Comment:
@vapourismo This is interesting. The instance is essentially the `Applicative` instance for `Const`; which makes sense since `ProductB` is very similar to `Applicative`. However, with the laws in its current form, there is no room for side-effects, so the instance becomes unlawful:

```haskell
(\(Pair a _) -> a) `bmap` (Const [1] `bprod` Const [2])
  = Const [1,2]
 /= Const [1]

const (buniq x) (Const [1])
  = Const []
 /= Const [1]
  = const x `bmap` Const [1]
```

It would be interesting to rethink the laws of this class to see if effects can be accommodated, which would bring it closer to `Applicative`.

Reply:
Yes, that makes sense. I'll remove it.